### ISSUE TITLE
Explains new version scheme for Nokee

### ISSUE THREAD

**Issue by lacasseio** [nokeedev/gradle-native#350]
The previous scheme, i.e. `<next-version>.<git-hash>` was enforced by Bintray (e.g. not allowing timestamp in the version). When Bintray announce it was retiring, we moved all our snapshots into the S3 bucket and all released versions into Maven Central. We could then choose whatever version scheme we wanted for snapshots. The next version + git hash scheme had issues when it comes to version ordering because of the git hash. We decided to use the version suffix `-<time-stamp>.<git-hash>` to allow for ordering as well as quickly know when-ish the version was published. We keep the git-hash so that users can quickly reference what is "included" in the published version.

As for the next-version aspect, we are moving toward a "live at head" philosophy with continuous delivery. The general idea we settled on is a forward rolling version using incremental patch version based on the git commit distance since the last official release. The minor/major versions will always be cut manually while the patch version would be cut automatically if all test passes.

**Comment by bric3** [comment]
That would be nice to have new version (release) announced on the github repository as well.

**Comment by lacasseio** [comment]
You are right, the next version is still going to be 0.5. We went into a bit of a rabbit hole regarding the configuration model and we hope to put some blog posts soon on what happens around that. The versions containing hashes are as good as an official release which are mostly the releases that happen so far. When we will move to a more "live at head" release cadence, patch versions will be automatic and without GitHub release tags. Minor and major versions will be announced and should be more frequent as well. We are currently cleaning up some code debt so we can release 0.5.

**Comment by bric3** [comment]
Thanks for the heads up !

**Comment by pedrolamarao** [comment]
I have recently considered the problem of version ordering while reviewing methodology and considered the solution proposed here. I think it does solve the problem of ordering, but, it diminishes build reproducibility. After reviewing a lot of posts around the web, I think using `git rev-list` is a good alternative which preserved build reproducibility. My team will start experimenting with build tags based on `git branch --show-current + `git rev-list --count --first-parent HEAD`. Maybe this is an interesting option for Nokee.

**Comment by lacasseio** [comment]
Thanks for the suggestion. I agree that build reproducibility is important. In theory, for the same source, the resulting JARs should be the same. We want to avoid any changing information in the final JAR such as build timestamp. We have a separate internal artifact that include some of the build information for each version which include the build timestamp. I think it's a good compromise which allow more information on-demand for a particular version. We want to promote to help users answer those kind of questions.

I agree that a self-contained build marker should be included in the distribution JAR. There are still some open question regarding ad-hoc snapshot released from different branches (non-master) including dirty branch. I believe we do include the branch name in the version artifact. As for dirty branch, we could include a patch of the local changes in the version artifact to complement the Git hash. There is still the issue of Git hashes only available locally.
